File: driver-scala/src/integration/scala/org/mongodb/scala/syncadapter/SyncClientSession.scala
/*
 * Copyright 2008-present MongoDB, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.mongodb.scala.syncadapter

import com.mongodb.{ ClientSessionOptions, MongoInterruptedException, ServerAddress, TransactionOptions }
import com.mongodb.client.{ ClientSession => JClientSession, TransactionBody }
import com.mongodb.session.ServerSession
import org.bson.{ BsonDocument, BsonTimestamp }
import org.mongodb.scala._
import com.mongodb.reactivestreams.client.syncadapter.{ SyncMongoClient => JSyncMongoClient }

case class SyncClientSession(wrapped: ClientSession, originator: Object) extends JClientSession {

  override def getPinnedServerAddress: ServerAddress = wrapped.getPinnedServerAddress

  override def getRecoveryToken: BsonDocument = wrapped.getRecoveryToken

  override def setRecoveryToken(recoveryToken: BsonDocument): Unit = wrapped.setRecoveryToken(recoveryToken)

  override def getOptions: ClientSessionOptions = wrapped.getOptions

  override def isCausallyConsistent: Boolean = wrapped.isCausallyConsistent

  override def getOriginator: Object = originator

  override def getServerSession: ServerSession = wrapped.getServerSession

  override def getOperationTime: BsonTimestamp = wrapped.getOperationTime

  override def advanceOperationTime(operationTime: BsonTimestamp): Unit = wrapped.advanceOperationTime(operationTime)

  override def advanceClusterTime(clusterTime: BsonDocument): Unit = wrapped.advanceClusterTime(clusterTime)

  override def getClusterTime: BsonDocument = wrapped.getClusterTime

  override def setSnapshotTimestamp(snapshotTimestamp: BsonTimestamp): Unit =
    wrapped.setSnapshotTimestamp(snapshotTimestamp)

  override def getSnapshotTimestamp: BsonTimestamp = wrapped.getSnapshotTimestamp

  override def close(): Unit = {
    wrapped.close()
    sleep(JSyncMongoClient.getSleepAfterSessionClose)
  }

  override def hasActiveTransaction: Boolean = wrapped.hasActiveTransaction

  override def notifyMessageSent: Boolean = wrapped.notifyMessageSent

  override def notifyOperationInitiated(operation: Object): Unit = wrapped.notifyOperationInitiated(operation)

  override def getTransactionOptions: TransactionOptions = wrapped.getTransactionOptions

  override def startTransaction(): Unit = wrapped.startTransaction()

  override def startTransaction(transactionOptions: TransactionOptions): Unit =
    wrapped.startTransaction(transactionOptions)

  override def commitTransaction(): Unit = wrapped.commitTransaction().toSingle().toFuture().get()

  override def abortTransaction(): Unit = wrapped.abortTransaction().toSingle().toFuture().get()

  override def withTransaction[T](transactionBody: TransactionBody[T]) = throw new UnsupportedOperationException

  override def withTransaction[T](transactionBody: TransactionBody[T], options: TransactionOptions) =
    throw new UnsupportedOperationException

  override def getTransactionContext: AnyRef = wrapped.getTransactionContext

  override def setTransactionContext(address: ServerAddress, transactionContext: Any): Unit =
    wrapped.setTransactionContext(address, transactionContext)

  override def clearTransactionContext(): Unit = wrapped.clearTransactionContext()

  private def sleep(millis: Long): Unit = {
    try Thread.sleep(millis)
    catch {
      case e: InterruptedException =>
        throw new MongoInterruptedException(null, e)
    }
  }
}

File: driver-kotlin-sync/src/integration/kotlin/com/mongodb/kotlin/client/syncadapter/SyncClientSession.kt
/*
 * Copyright 2008-present MongoDB, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.mongodb.kotlin.client.syncadapter

import com.mongodb.ClientSessionOptions
import com.mongodb.ServerAddress
import com.mongodb.TransactionOptions
import com.mongodb.client.ClientSession as JClientSession
import com.mongodb.client.TransactionBody
import com.mongodb.kotlin.client.ClientSession
import com.mongodb.session.ServerSession
import org.bson.BsonDocument
import org.bson.BsonTimestamp

internal class SyncClientSession(internal val wrapped: ClientSession, private val originator: Any) : JClientSession {
    private val delegate: JClientSession = wrapped.wrapped

    override fun close(): Unit = delegate.close()

    override fun getPinnedServerAddress(): ServerAddress? = delegate.pinnedServerAddress

    override fun getTransactionContext(): Any? = delegate.transactionContext

    override fun setTransactionContext(address: ServerAddress, transactionContext: Any): Unit =
        delegate.setTransactionContext(address, transactionContext)

    override fun clearTransactionContext(): Unit = delegate.clearTransactionContext()

    override fun getRecoveryToken(): BsonDocument? = delegate.recoveryToken

    override fun setRecoveryToken(recoveryToken: BsonDocument): Unit {
        delegate.recoveryToken = recoveryToken
    }

    override fun getOptions(): ClientSessionOptions = delegate.options

    override fun isCausallyConsistent(): Boolean = delegate.isCausallyConsistent

    override fun getOriginator(): Any = originator

    override fun getServerSession(): ServerSession = delegate.serverSession

    override fun getOperationTime(): BsonTimestamp = delegate.operationTime

    override fun advanceOperationTime(operationTime: BsonTimestamp?): Unit =
        delegate.advanceOperationTime(operationTime)

    override fun advanceClusterTime(clusterTime: BsonDocument?): Unit = delegate.advanceClusterTime(clusterTime)

    override fun setSnapshotTimestamp(snapshotTimestamp: BsonTimestamp?) {
        delegate.snapshotTimestamp = snapshotTimestamp
    }

    override fun getSnapshotTimestamp(): BsonTimestamp? = delegate.snapshotTimestamp

    override fun getClusterTime(): BsonDocument = delegate.clusterTime

    override fun hasActiveTransaction(): Boolean = delegate.hasActiveTransaction()

    override fun notifyMessageSent(): Boolean = delegate.notifyMessageSent()

    override fun notifyOperationInitiated(operation: Any): Unit = delegate.notifyOperationInitiated(operation)

    override fun getTransactionOptions(): TransactionOptions = delegate.transactionOptions

    override fun startTransaction(): Unit = delegate.startTransaction()

    override fun startTransaction(transactionOptions: TransactionOptions): Unit =
        delegate.startTransaction(transactionOptions)

    override fun commitTransaction(): Unit = delegate.commitTransaction()

    override fun abortTransaction(): Unit = delegate.abortTransaction()

    override fun <T : Any> withTransaction(transactionBody: TransactionBody<T>): T =
        throw UnsupportedOperationException()

    override fun <T : Any> withTransaction(transactionBody: TransactionBody<T>, options: TransactionOptions): T =
        throw UnsupportedOperationException()
}
